File: code_examples/DesignPatterns/decorator.kt
interface caterpillar {
    fun eat(): String
}

class BasicCaterpillar : caterpillar {
    override fun eat(): String {
        return "The Very Hungry Caterpillar is eating"
    }
}

class PearDecorator(private val caterpillar: caterpillar) : caterpillar {
    override fun eat(): String {
        return "${caterpillar.eat()} one pear"
    }
}

class AppleDecorator(private val caterpillar: caterpillar) : caterpillar {
    override fun eat(): String {
        return "${caterpillar.eat()} two apples"
    }
}

class BananaDecorator(private val caterpillar: caterpillar) : caterpillar {
    override fun eat(): String {
        return "${caterpillar.eat()} three bananas"
    }
}

fun main () {
    val basicCaterpillar = BasicCaterpillar()
    val pearCaterpillar = PearDecorator(basicCaterpillar)
    val appleCaterpillar = AppleDecorator(basicCaterpillar)
    val bananaCaterpillar = BananaDecorator(basicCaterpillar)
    val bananaAndAppleCaterpillar = BananaDecorator(PearDecorator(basicCaterpillar))

    println(basicCaterpillar.eat())
    println(pearCaterpillar.eat())
    println(appleCaterpillar.eat())
    println(bananaCaterpillar.eat())
    println(bananaAndAppleCaterpillar.eat())
}

File: code_examples/DesignPatterns/Decorator.js
class VeryHungryCaterpillar {
    eat() {
        return "The Very Hungry Caterpillar is eating";
    }
}

class BasicCaterpillar extends VeryHungryCaterpillar {}

class PearDecorator extends BasicCaterpillar {
    constructor(caterpillar) {
        super();
        this.caterpillar = caterpillar;
    }

    eat() {
        return `${this.caterpillar.eat()}  one pear.`;
    }
}

class AppleDecorator extends BasicCaterpillar {
    constructor(caterpillar) {
        super();
        this.caterpillar = caterpillar;
    }

    eat() {
        return `${this.caterpillar.eat()} two apples.`;
    }
}

class BananaDecorator extends BasicCaterpillar {
    constructor(caterpillar) {
        super();
        this.caterpillar = caterpillar;
    }

    eat() {
        return `${this.caterpillar.eat()} three bananas.`;
    }
}

const basicCaterpillar = new BasicCaterpillar();
const pearCaterpillar = new PearDecorator(basicCaterpillar);
const appleCaterpillar = new AppleDecorator(basicCaterpillar);
const bananaCaterpillar = new BananaDecorator(basicCaterpillar);
const bananaAndAppleCaterpillar = new BananaDecorator(new AppleDecorator(basicCaterpillar));

console.log(basicCaterpillar.eat());
console.log(pearCaterpillar.eat());
console.log(appleCaterpillar.eat());
console.log(bananaCaterpillar.eat());
console.log(bananaAndAppleCaterpillar.eat());
